File: erettsegi/4_Futar/futar.pas
//Emelt érettségi 2012 május, https://www.oktatas.hu/pub_bin/dload/kozoktatas/erettsegi/feladatok2012tavasz/emelt/e_inf_12maj_fl.pdf
program futar;
uses crt, strutils, sysutils;

const adatok = 'tavok.txt';
	  max = 61;

Type Jegyzet = record
	Nap:integer;
	Fuvar:integer;
	Km:integer;
END;

var Lista:array[1..max] of Jegyzet;
	fajl:text;
	i:integer;
	k:integer;
	g:integer;
	ossz:integer;

Procedure Beolvasas();
var sor:string;
	darab:array[1..3] of string;
BEGIN
	Assign(fajl, adatok);
	Reset(fajl);
	
	While NOT EOF(fajl) do BEGIN
		ReadLn(fajl, sor);
		for i := 1 to 3 do BEGIN
			darab[i] := ExtractWord(i, sor, [' ']);
		END;
		
		ossz := ossz + 1;
		Lista[ossz].Nap := StrToInt(darab[1]);
		Lista[ossz].Fuvar := StrToInt(darab[2]);
		Lista[ossz].Km := StrToInt(darab[3]);
	END;
	
	Close(fajl);
END;

Procedure Feladat2();
var index:integer = 1;
	lnap:integer = 8;
	osszesen:integer;
BEGIN
	WriteLn('2. feladat:');
	
	for i := 1 to ossz do BEGIN
		if(Lista[i].Nap < lnap) then BEGIN
			lnap := Lista[i].Nap;
		END;
	END;
	
	osszesen := ossz;
	While NOT(osszesen = 0) do BEGIN
		if(Lista[osszesen].Nap = lnap) then BEGIN
			if(Lista[osszesen].Fuvar < Lista[index].Fuvar) then BEGIN
				index := osszesen;
			END;
		END;
		
		osszesen := osszesen - 1;
	END;
	
	WriteLn('A hét legelső útja ', Lista[index].Km, ' km volt!');
END;

Procedure Feladat3();
var index:integer = 1;
	lnap:integer = 0;
	osszesen:integer;
BEGIN
	WriteLn('3. feladat:');
	
	for i := 1 to ossz do BEGIN
		if(Lista[i].Nap > lnap) then BEGIN
			lnap := Lista[i].Nap;
		END;
	END;
	
	osszesen := ossz;
	While NOT(osszesen = 0) do BEGIN
		if(Lista[osszesen].Nap = lnap) then BEGIN
			if(Lista[osszesen].Fuvar > Lista[index].Fuvar) then BEGIN
				index := osszesen;
			END;
		END;
		
		osszesen := osszesen - 1;
	END;
	
	WriteLn('A hét utolsó útja ', Lista[index].Km, ' km volt!');
END;

Procedure Feladat4();
var dolgozott:array[1..7] of boolean;
BEGIN
	for i := 1 to 7 do BEGIN
		dolgozott[i] := false;
	END;
	
	for i := 1 to ossz do BEGIN
		dolgozott[Lista[i].Nap] := true;
	END;
	
	Write('A futár ezeken a napokon nem dolgozott: ');
	for i := 1 to 7 do BEGIN
		if(dolgozott[i] = false) then BEGIN
			Write(i, ' ');
		END;
	END;
END;

Procedure Feladat5();
var fuvarok:array[1..7] of integer;
	nap:integer = 1;
BEGIN
	WriteLn('5. feladat: ');
	
	for i := 1 to 7 do BEGIN
		fuvarok[i] := 0;
	END;
	
	for i := 1 to ossz do BEGIN
		fuvarok[Lista[i].Nap] := fuvarok[Lista[i].Nap] + 1;
	END;
	
	for i := 1 to 7 do BEGIN
		if(fuvarok[i] > fuvarok[nap]) then BEGIN
			nap := i;
		END;
	END;
	
	WriteLn('A(z) ', nap, '. napon volt a legtöbb fuvar.');
END;

Procedure Feladat6();
var ut:array[1..7] of integer;
BEGIN
	WriteLn('6. feladat: ');
	
	for i := 1 to 7 do BEGIN
		ut[i] := 0;
	END;
	
	for i := 1 to ossz do BEGIN
		ut[Lista[i].Nap] := ut[Lista[i].Nap]+Lista[i].Km;
	END;
	
	for i := 1 to 7 do BEGIN
		WriteLn(i, '. nap: ', ut[i],' km ');
	END;
END;

Function Fuvardij(km:integer): integer;
var dij:integer;
BEGIN
	Case km Of
		1..2:dij := 500;
		3..5:dij := 700;
		6..10:dij := 900;
		11..20:dij := 1400;
		21..30:dij := 2000;
	END;
	
	Fuvardij := dij;
END;

Procedure Feladat7();
var km:integer;
BEGIN
	WriteLn('7. feladat: ');
	Write('Adj meg egy tetszőleges távolságot: ');
	ReadLn(km);
		
	WriteLn('Díjazás: ', Fuvardij(km));
END;

Function HanyFuvar(nap:integer): integer;
var j:integer;
	db:integer = 0;
BEGIN
	for j := 1 to ossz do BEGIN
		if(Lista[j].Nap = nap) then BEGIN
			db := db + 1;
		END;
	END;
	
	HanyFuvar := db;
END;

Procedure Feladat8();
BEGIN
	Assign(fajl, 'dijazas.txt');
	ReWrite(fajl);
	
	for k := 1 to 7 do BEGIN
		for g := 1 to HanyFuvar(k) do BEGIN
			for i := 1 to ossz do BEGIN
				if((Lista[i].Nap = k) AND (Lista[i].Fuvar = g)) then BEGIN
					WriteLn(fajl, Lista[i].Nap, '. nap ', g, '. út: ', Fuvardij(Lista[i].Km), ' Ft');
				END;
			END;
		END;
	END;
	
	Close(fajl);
END;

Procedure Feladat9();
var fizetes:int64 = 0;
BEGIN
	WriteLn('9. feladat: ');
	
	for k := 1 to 7 do BEGIN
		for g := 1 to HanyFuvar(k) do BEGIN
			for i := 1 to ossz do BEGIN
				if((Lista[i].Nap = k) AND (Lista[i].Fuvar = g)) then BEGIN
					fizetes := fizetes + Fuvardij(Lista[i].Km);
				END;
			END;
		END;
	END;
	
	WriteLn('A futár ', fizetes, ' Ft fizetést kap a heti munkájáért!');
END;

BEGIN
	ClrScr;
	
	Beolvasas();
	Feladat2();
	WriteLn;
	Feladat3();
	WriteLn;
	Feladat4();
	WriteLn;
	WriteLn;
	Feladat5();
	WriteLn;
	Feladat6();
	WriteLn;
	Feladat7();
	WriteLn;
	Feladat8();
	WriteLn;
	Feladat9();
	
	WriteLn;
	WriteLn;
END.

File: erettsegi/4_Futar/futar.cs
using System;
using System.IO;
using System.Linq;
using System.Collections.Generic;

namespace tomikaasd
{
    class futar
    {
        class Fuvar
        {
            public int Nap { get; set; }
            public int Hanyadik { get; set; }
            public int Tavolsag { get; set; }

            public Fuvar() { }

            public Fuvar(int nap, int hanyadik, int tavolsag)
            {
                Nap = nap;
                Hanyadik = hanyadik;
                Tavolsag = tavolsag;
            }
        }

        static List<Fuvar> Fuvarok = new List<Fuvar>();

        static void Main(string[] args)
        {
            Feladat1();
            Feladat2();
            Console.WriteLine();
            Feladat3();
            Console.WriteLine();
            Feladat4();
            Console.WriteLine();
            Feladat5();
            Console.WriteLine();
            Feladat6();
            Console.WriteLine();
            Feladat7();
            Console.WriteLine();
            Feladat8();
            Feladat9();
        }

        static void Feladat1()
        {
            FileStream fs = File.Open("tavok.txt", FileMode.Open, FileAccess.Read);
            StreamReader sr = new StreamReader(fs);

            while (!sr.EndOfStream)
            {
                string[] darab = sr.ReadLine().Split(' ');
                Fuvarok.Add(new Fuvar(Convert.ToInt32(darab[0]), Convert.ToInt32(darab[1]), Convert.ToInt32(darab[2])));
            }

            sr.Close();
            fs.Close();
        }

        static void Feladat2()
        {
            Console.WriteLine("2. feladat:");

            int legkisebbNap = Fuvarok[0].Nap;
            foreach(Fuvar i in Fuvarok)
            {
                if (i.Nap < legkisebbNap)
                {
                    legkisebbNap = i.Nap;
                }
            }

            Fuvar legelsoFuvar = Fuvarok[0];
            for(int i = 1; i < Fuvarok.Count; i++)
            {
                if (Fuvarok[i].Nap != legkisebbNap)
                    continue;

                if(Fuvarok[i].Hanyadik < legelsoFuvar.Hanyadik)
                {
                    legelsoFuvar = Fuvarok[i];
                }
            }

            Console.WriteLine($"A hét legelső útja {legelsoFuvar.Nap}. nap, {legelsoFuvar.Tavolsag} km");
        }

        static void Feladat3()
        {
            Console.WriteLine("3. feladat:");

            int legnagyobbNap = Fuvarok[0].Nap;
            foreach (Fuvar i in Fuvarok)
            {
                if (i.Nap > legnagyobbNap)
                {
                    legnagyobbNap = i.Nap;
                }
            }

            Fuvar legutolsoFuvar = Fuvarok[0];
            for (int i = 1; i < Fuvarok.Count; i++)
            {
                if (Fuvarok[i].Nap != legnagyobbNap)
                    continue;

                if (Fuvarok[i].Hanyadik > legutolsoFuvar.Hanyadik)
                {
                    legutolsoFuvar = Fuvarok[i];
                }
            }

            Console.WriteLine($"A hét legutolsó útja {legutolsoFuvar.Nap}. nap, {legutolsoFuvar.Tavolsag} km");
        }

        static void Feladat4()
        {
            Console.WriteLine("4. feladat:");

            bool[] dolgozott = new bool[8];
            foreach(Fuvar i in Fuvarok)
            {
                dolgozott[i.Nap] = true;
            }

            for(int i = 1; i < 8; i++)
            {
                if(!dolgozott[i])
                {
                    Console.WriteLine($"A(z) {i}. nap szabadnap volt!");
                }
            }
        }

        static void Feladat5()
        {
            Console.WriteLine("5. feladat:");

            int[] fuvarDb = new int[8];
            foreach (Fuvar i in Fuvarok)
            {
                fuvarDb[i.Nap] = 0;
            }

            foreach(Fuvar i in Fuvarok)
            {
                fuvarDb[i.Nap]++;
            }

            for(int i = 1; i < 8; i++)
            {
                if (fuvarDb[i] == 0)
                    continue;

                Console.WriteLine($"{i}. napon {fuvarDb[i]} fuvar volt.");
            }
        }

        static void Feladat6()
        {
            Console.WriteLine("6. feladat:");

            int[] fuvarKm = new int[8];
            foreach (Fuvar i in Fuvarok)
            {
                fuvarKm[i.Nap] = 0;
            }

            /*
            for(int i = 0; i < Fuvarok.Count; i++)
            {
                fuvarKm[Fuvarok[i].Nap] = 0;
            }
            */

            foreach (Fuvar i in Fuvarok)
            {
                fuvarKm[i.Nap] += i.Tavolsag;
            }

            /*
            for(int i = 0; i < Fuvarok.Count; i++)
            {
                fuvarKm[Fuvarok[i].Nap] += Fuvarok[i].Tavolsag;
            }
            */

            for (int i = 1; i < 8; i++)
            {
                if (fuvarKm[i] == 0)
                    continue;

                Console.WriteLine($"{i}. nap: {fuvarKm[i]} km");
            }
        }

        static void Feladat7()
        {
            Console.WriteLine("7. feladat:");

            Console.Write("Adj meg egy kilometert: ");
            int km = Convert.ToInt32(Console.ReadLine());
            Console.WriteLine($"A futár fizetése {km} km-ért {Fizetes(km)} Ft.");
        }

        static void Feladat8()
        {
            FileStream fs = File.Open("dijazas.txt", FileMode.Create, FileAccess.Write);
            StreamWriter sw = new StreamWriter(fs);

            Fuvarok = Fuvarok.OrderBy(x => x.Hanyadik).ToList();

            for(int i = 1; i < 8; i++)
            {
                foreach (Fuvar fuvar in Fuvarok)
                {
                    if (fuvar.Nap != i)
                        continue;

                    sw.WriteLine($"{i}. nap {fuvar.Hanyadik}. út: {Fizetes(fuvar.Tavolsag)}");
                }
            }

            sw.Close();
            fs.Close();
        }

        static void Feladat9()
        {
            Console.WriteLine("9. feladat:");
            int fizetes = 0;

            foreach(Fuvar i in Fuvarok)
            {
                fizetes += Fizetes(i.Tavolsag);
            }

            Console.WriteLine($"A futár a heti munkájáért {fizetes} Ft-t kap.");
        }

        static int Fizetes(int km)
        {
            if (km < 2) return 500;
            else if (km < 5) return 700;
            else if (km < 10) return 900;
            else if (km < 20) return 1400;
            
            return 2000;
        }
    }
}